beacon_node/eth2_libp2p/src/peer_manager/mod.rs

```diff
@@ -214,18 +214,63 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
 
     /// A request to find peers on a given subnet.
     pub fn discover_subnet_peers(&mut self, subnets_to_discover: Vec<SubnetDiscovery>) {
-        // Extend the time to maintain peers if required.
-        for s in subnets_to_discover.iter() {
-            if let Some(min_ttl) = s.min_ttl {
-                self.network_globals
+        let filtered: Vec<SubnetDiscovery> = subnets_to_discover
+            .into_iter()
+            .filter(|s| {
+                // Extend min_ttl of connected peers on required subnets
+                if let Some(min_ttl) = s.min_ttl {
+                    self.network_globals
+                        .peers
+                        .write()
+                        .extend_peers_on_subnet(s.subnet_id, min_ttl);
+                }
+                // Already have target number of peers, no need for subnet discovery
+                let peers_on_subnet = self
+                    .network_globals
                     .peers
-                    .write()
-                    .extend_peers_on_subnet(s.subnet_id, min_ttl);
-            }
-        }
+                    .read()
+                    .peers_on_subnet(s.subnet_id)
+                    .count();
+                if peers_on_subnet >= TARGET_SUBNET_PEERS {
+                    debug!(
+                        self.log,
+                        "Discovery query ignored";
+                        "subnet_id" => format!("{:?}",s.subnet_id),
+                        "reason" => "Already connected to desired peers",
+                        "connected_peers_on_subnet" => peers_on_subnet,
+                        "target_subnet_peers" => TARGET_SUBNET_PEERS,
+                    );
+                    false
+                // No need for costly discovery query if we find required peers in cached enrs
+                // Attempt to dial the cached enrs in discovery for peers in subnet id
+                // and check if we got required number of peers.
+                } else {
+                    self.dial_cached_enrs_in_subnet(s.subnet_id);
+                    let peers_on_subnet = self
+                        .network_globals
+                        .peers
+                        .read()
+                        .peers_on_subnet(s.subnet_id)
+                        .count();
+                    if peers_on_subnet >= TARGET_SUBNET_PEERS {
```

Comment:
The count isn't going to change here. The peers will likey just shift to `dialing` state and won't be coutned. So this check seems like it will never hit.

Reply:
I think its fine to just to `dial_cached_enrs_in_subnet` (these are likely stale) and then send off the discovery query. 
It will get queued. If we manage to connect by the time the subnet query starts, it will do this check for us.